File: python_packages/habana_frameworks/torch/hpex/kernels/FusedSDPA.py
###############################################################################
#
#  Copyright (c) 2021-2025 Intel Corporation
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
###############################################################################

import math  # for sqrt etc
import os

import habana_frameworks.torch.hpu as ht

import torch


# Please refer to FusedSDPA documentation at:
# https://docs.habana.ai/en/latest/PyTorch/Python_Packages.html#hpex-kernels-fusedsdpa
def check_dbg_env_var(v):
    env_var_set = False
    if int(os.getenv(v, 0)) == 1:
        env_var_set = True
    return env_var_set


def is_gqa(q, k):
    gqa = False
    dims = q.dim()
    if dims == 4:
        q_heads = q.shape[1]
        kv_heads = k.shape[1]
        gqa = (q_heads != kv_heads) and kv_heads != 1
    return gqa


def gqa_input_reshape_bwd(q, v, grad_in):
    new_shape = (q.shape[0], q.shape[1], q.shape[2], q.shape[3], v.shape[-1])
    return grad_in.reshape(new_shape)


def gqa_input_reshape_fwd(q, k, v, attention_mask):
    q_heads = q.shape[1]
    kv_heads = k.shape[1]

    q_heads_per_group = q_heads // kv_heads
    groups = kv_heads

    bs, heads, seq_len, h_dim = q.shape
    new_q_shape = (bs, groups, q_heads_per_group, seq_len, h_dim)
    q = q.reshape(new_q_shape)

    bs, heads, seq_len, h_dim = k.shape
    new_k_shape = (bs, groups, 1, seq_len, h_dim)
    k = k.reshape(new_k_shape)

    bs, heads, seq_len, h_dim = v.shape
    new_v_shape = (bs, groups, 1, seq_len, h_dim)
    v = v.reshape(new_v_shape)

    if attention_mask is not None:
        bs, heads, seq_len_t, seq_len_s = attention_mask.shape
        if heads == q_heads:  # attention mask shape = [batch size, q_heads, *, *]
            new_attn_mask_shape = (bs, groups, q_heads_per_group, seq_len_t, seq_len_s)
            attention_mask = attention_mask.reshape(new_attn_mask_shape)
        else:  # attention mask shape = [batch size, 1, *, *]
            attention_mask = attention_mask.unsqueeze(1)  # add groups dim and set to 1

    return q, k, v, attention_mask


def gqa_output_reshape(tensor):
    bs, groups, heads_per_group, seq_len, h_dim = tensor.shape
    new_shape = (bs, groups * heads_per_group, seq_len, h_dim)
    return tensor.reshape(new_shape)


def sdpa_fwd_wrapper(
    ctx,
    q,
    k,
    v,
    attn_mask=None,
    dropout_p=0.0,
    is_causal=False,
    scale=None,
    softmax_mode="None",
    recompute_mode=None,
    valid_seq_len=None,
    seq_padding_type="left",
    return_dropout_mask=False,
    return_attn_probs=False,
):
    requires_backward = q.requires_grad or k.requires_grad or v.requires_grad

    # Handle zero sized tensors(for now only in inference) by returning a dummy output.
    if requires_backward is False:
        if q.numel() == 0 or k.numel() == 0 or v.numel() == 0:
            out_shape = list(q.shape)
            out_shape[-1] = v.shape[-1]
            dummy_out = q.new_empty(out_shape, requires_grad=requires_backward, layout=q.layout)
            return dummy_out

    softmax_mode = softmax_mode.lower()
    seq_padding_type = seq_padding_type.lower()
    if scale is None:
        scale = 1.0 / math.sqrt(q.size(-1))

    # Check if recompute variant is enabled
    recompute = recompute_mode
    if recompute is None:
        recompute = ht.recompute_sdp_enabled()

    q_seq_len = q.size(-2)
    # In case of inference, override the mode set by user and set the mode internally
    # and go via recmpute mode unless returing attn prob is requested (since attn prob
    # is returned only in non-recomp mode). For now take recomp path only in case Q seq len is 1.
    # Later see if this condition can be removed.
    if requires_backward is False:
        if q_seq_len == 1:
            recompute = True
        # force recomp to False if attn prob is to be returned
        if return_attn_probs:
            recompute = False

    if return_attn_probs is True:
        assert requires_backward is False, "return_attn_probs is supported only for inference mode"

    if recompute and requires_backward and softmax_mode == "fast":
        assert (
            is_causal
        ), "Optimized softmax mode is supported in recompute training mode only in causal(triangular) mask case"

    if valid_seq_len is not None:
        assert is_causal and (
            requires_backward is False
        ), "Valid sequence length is supported only in inference with is_causal(triangular) mask case"

    if recompute:
        assert return_dropout_mask is False, "Return_dropout_mask is not supported in recompute mode"

    if softmax_mode == "fp32":
        q_dtype = q.dtype
        assert (
            requires_backward is False and q_dtype == torch.bfloat16
        ), "softmax_mode = fp32 is supported only for inference mode and when q/k/v inputs are BF16"

    gqa = is_gqa(q, k)
    if gqa:
        q, k, v, attn_mask = gqa_input_reshape_fwd(q, k, v, attn_mask)
    if recompute:
        out, m, linv, seed = torch.ops.hpu.sdpa_recomp_fwd(
            q,
            k,
            v,
            attn_mask,
            dropout_p,
            scale,
            is_causal,
            requires_backward,
            softmax_mode,
            valid_seq_len,
            seq_padding_type,
        )
        if gqa:
            out = gqa_output_reshape(out)
        if not requires_backward:
            return out
        ctx.save_for_backward(q, k, v, attn_mask, m, linv, seed, out)
    else:
        out, P, dm = torch.ops.hpu.sdpa_fwd(
            q, k, v, attn_mask, dropout_p, scale, is_causal, softmax_mode, valid_seq_len, seq_padding_type
        )
        if gqa:
            out = gqa_output_reshape(out)
        if not requires_backward:
            # return_dropout_mask   return_attn_probs       Outputs
            #       FALSE               FALSE               fwd_out
            #       FALSE               TRUE                fwd_out, P
            #       TRUE                FALSE               fwd_out, dm
            #       TRUE                TRUE                fwd_out, P, dm
            if gqa:
                if dropout_p > 0.0:
                    dm = gqa_output_reshape(dm)
                if return_attn_probs:
                    P = gqa_output_reshape(P)

            if return_dropout_mask:
                if return_attn_probs:
                    return out, P, dm
                else:
                    return out, dm
            else:
                if return_attn_probs:
                    return out, P
                else:
                    return out

        ctx.save_for_backward(q, k, v, P, dm, out)

    ctx.dropout_p = dropout_p
    ctx.scale = scale
    ctx.is_causal = is_causal
    ctx.recompute = recompute
    ctx.gqa = gqa
    ctx.softmax_mode = softmax_mode

    if recompute:
        return out

    if not return_dropout_mask:
        return out
    else:
        if dropout_p > 0.0 and gqa:
            dm = gqa_output_reshape(dm)
        return out, dm


def sdpa_bwd_wrapper(ctx, dout, *args):
    if ctx.recompute:
        q, k, v, attn_mask, m, linv, seed, fwd_out = ctx.saved_tensors
        scale = ctx.scale
        dropout_p = ctx.dropout_p
        is_causal = ctx.is_causal
        softmax_mode = ctx.softmax_mode
        if ctx.gqa:
            dout = gqa_input_reshape_bwd(q, v, dout)
            fwd_out = gqa_input_reshape_bwd(q, v, fwd_out)
        dq, dk, dv = torch.ops.hpu.sdpa_recomp_bwd(
            dout, q, k, v, attn_mask, m, linv, seed, is_causal, dropout_p, scale, softmax_mode, fwd_out
        )
        if ctx.gqa:
            dq = gqa_output_reshape(dq)
            dk = gqa_output_reshape(dk)
            dv = gqa_output_reshape(dv)
        return dq, dk, dv, None, None, None, None, None, None, None, None, None, None
    else:
        q, k, v, P, dm, fwd_out = ctx.saved_tensors
        scale = ctx.scale
        is_causal = ctx.is_causal
        dropout_p = ctx.dropout_p
        if ctx.gqa:
            dout = gqa_input_reshape_bwd(q, v, dout)
            fwd_out = gqa_input_reshape_bwd(q, v, fwd_out)
        dq, dk, dv = torch.ops.hpu.sdpa_bwd(dout, q, k, v, P, dm, is_causal, dropout_p, scale, fwd_out)
        if ctx.gqa:
            dq = gqa_output_reshape(dq)
            dk = gqa_output_reshape(dk)
            dv = gqa_output_reshape(dv)
        return dq, dk, dv, None, None, None, None, None, None, None, None, None, None


class FusedSDPA(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx,
        q,
        k,
        v,
        attn_mask=None,
        dropout_p=0.0,
        is_causal=False,
        scale=None,
        softmax_mode="None",
        recompute_mode=None,
        valid_seq_len=None,
        seq_padding_type="left",
        return_dropout_mask=False,
        return_attn_probs=False,
    ):
        return sdpa_fwd_wrapper(
            ctx,
            q,
            k,
            v,
            attn_mask=attn_mask,
            dropout_p=dropout_p,
            is_causal=is_causal,
            scale=scale,
            softmax_mode=softmax_mode,
            recompute_mode=recompute_mode,
            valid_seq_len=valid_seq_len,
            seq_padding_type=seq_padding_type,
            return_dropout_mask=return_dropout_mask,
            return_attn_probs=return_attn_probs,
        )

    @staticmethod
    def backward(ctx, dout, *args):
        return sdpa_bwd_wrapper(ctx, dout, *args)

File: habana_eager/ops/fused_sdpa.cpp
/**
 * Copyright (c) 2025 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "fused_sdpa.h"
#include <ATen/native/transformers/sdp_utils_cpp.h>
#include "backend/random.h"
#include "common/dump_args.h"
#include "generated/backend/sdpa_bwd.h"
#include "habana_eager/ops/eager_op.h"
#include "habana_helpers/logging.h"
#include "hpu_ops/op_logger.h"
#include "hpu_ops/sdpa_gen.h"

namespace habana {
namespace eager {

int64_t fused_sdp_choice_hpu(
    [[maybe_unused]] const at::Tensor& query,
    [[maybe_unused]] const at::Tensor& key,
    [[maybe_unused]] const at::Tensor& value,
    [[maybe_unused]] const ::std::optional<at::Tensor>& attn_mask,
    [[maybe_unused]] double dropout_p,
    [[maybe_unused]] bool is_causal,
    [[maybe_unused]] ::std::optional<double> scale,
    [[maybe_unused]] bool enable_gqa) {
  // there are five availble SDPBackend (math, flash_attention,
  // efficient_attention, cudnn_attention, overrideable)
  return static_cast<int64_t>(sdp::SDPBackend::math);
}

std::tuple<at::Tensor, at::Tensor, at::Tensor> dispatch_sdpa_fwd_wrap(
    const at::Tensor& q,
    const at::Tensor& k,
    const at::Tensor& v,
    const std::optional<at::Tensor>& attention_mask,
    const double p,
    const double scale,
    const bool is_causal,
    std::string_view softmax_mode,
    const std::optional<at::Tensor>& valid_seq_len,
    std::string_view seq_padding_type) {
  PT_EAGER_TRACE;
  PT_OP_INFO(
      "sdpa_fwd :",
      DUMP_10ARGS(
          q,
          k,
          v,
          attention_mask,
          p,
          scale,
          is_causal,
          softmax_mode,
          valid_seq_len,
          seq_padding_type));
  static auto op = torch::Dispatcher::singleton()
                       .findSchemaOrThrow("hpu::sdpa_fwd", "")
                       .typed<decltype(dispatch_sdpa_fwd_wrap)>();
  return op.call(
      q,
      k,
      v,
      attention_mask,
      p,
      scale,
      is_causal,
      softmax_mode,
      valid_seq_len,
      seq_padding_type);
}

std::tuple<at::Tensor, at::Tensor, at::Tensor> dispatch_sdpa_bwd_wrap(
    const at::Tensor& grad,
    const at::Tensor& q,
    const at::Tensor& k,
    const at::Tensor& v,
    const at::Tensor& P,
    const std::optional<at::Tensor>& dm,
    const bool is_causal,
    const double p,
    const double scale,
    const at::Tensor& fwd_out) {
  PT_EAGER_TRACE;
  PT_OP_INFO(
      "sdpa_bwd :",
      DUMP_10ARGS(grad, q, k, v, P, dm, is_causal, p, scale, fwd_out));

  static auto op = torch::Dispatcher::singleton()
                       .findSchemaOrThrow("hpu::sdpa_bwd", "")
                       .typed<decltype(dispatch_sdpa_bwd_wrap)>();
  return op.call(grad, q, k, v, P, dm, is_causal, p, scale, fwd_out);
}

at::Tensor gqa_output_reshape(at::Tensor input_tensor) {
  auto size = input_tensor.sizes().vec();
  return input_tensor.reshape({size[0], size[1] * size[2], size[3], size[4]});
}

std::vector<at::Tensor> gqa_input_reshape_fwd(
    at::Tensor& query,
    at::Tensor& key,
    at::Tensor& value,
    ::std::optional<at::Tensor>& attn_mask) {
  auto q_size = query.sizes().vec();
  auto k_size = key.sizes().vec();
  auto v_size = value.sizes().vec();

  auto q_heads = q_size[1];
  auto kv_heads = k_size[1];

  auto q_heads_per_group = q_heads / kv_heads;
  auto groups = kv_heads;

  query = query.reshape(
      {q_size[0], groups, q_heads_per_group, q_size[2], q_size[3]});
  key = key.reshape({k_size[0], groups, 1, k_size[2], k_size[3]});
  value = value.reshape({v_size[0], groups, 1, v_size[2], v_size[3]});

  if (attn_mask.has_value()) {
    auto a_size = attn_mask.value().sizes().vec();
    if (q_heads ==
        a_size[1]) { // attention mask shape = [batch size, q_heads, *, *]
      attn_mask = attn_mask.value().reshape(
          {a_size[0], groups, q_heads_per_group, a_size[2], a_size[3]});
    } else { // attention mask shape = [batch size, 1, *, *]
      attn_mask = attn_mask.value().unsqueeze(1);
    }
  }
  std::vector out{query, key, value};
  out.push_back(attn_mask.has_value() ? attn_mask.value() : torch::Tensor());
  return out;
}

at::Tensor gqa_input_reshape_bwd(
    at::Tensor& query,
    at::Tensor& value,
    at::Tensor grad) {
  auto q_size = query.sizes().vec();
  auto v_size = value.sizes().vec();
  return grad.reshape({q_size[0], q_size[1], q_size[2], q_size[3], v_size[3]});
}

class FusedSDPAAutogradHPU
    : public torch::autograd::Function<FusedSDPAAutogradHPU> {
 public:
  static at::Tensor forward(
      torch::autograd::AutogradContext* ctx,
      const at::Tensor& query,
      const at::Tensor& key,
      const at::Tensor& value,
      const ::std::optional<at::Tensor>& attn_mask,
      double dropout_p,
      bool is_causal,
      ::std::optional<double> scale,
      bool enable_gqa) {
    PT_EAGER_TRACE;
    auto softmax_mode = "None";
    auto seq_padding_type = "left";
    double scale_;
    if (!scale.has_value())
      scale_ = 1 / sqrt(query.sizes()[3]);
    else
      scale_ = scale.value();
    auto valid_seq_len = std::optional<at::Tensor>();
    ctx->saved_data["dropout_p"] = dropout_p;
    ctx->saved_data["scale"] = scale_;
    ctx->saved_data["is_causal"] = is_causal;
    ctx->saved_data["enable_gqa"] = enable_gqa;

    at::Tensor query_n = query;
    at::Tensor key_n = key;
    at::Tensor value_n = value;
    ::std::optional<at::Tensor> attn_mask_n = attn_mask;
    if (enable_gqa) {
      auto gqa_out =
          gqa_input_reshape_fwd(query_n, key_n, value_n, attn_mask_n);
      query_n = gqa_out[0];
      key_n = gqa_out[1];
      value_n = gqa_out[2];
      attn_mask_n = attn_mask.has_value() ? gqa_out[3] : attn_mask_n;
    }

    // output (out, P, dm)
    auto output = dispatch_sdpa_fwd_wrap(
        query_n,
        key_n,
        value_n,
        attn_mask_n,
        dropout_p,
        scale_,
        is_causal,
        softmax_mode,
        valid_seq_len,
        seq_padding_type);
    auto out = std::get<0>(output);
    auto P = std::get<1>(output);
    auto dm = std::get<2>(output);
    if (enable_gqa) {
      out = gqa_output_reshape(out);
      P = gqa_output_reshape(P);
      if (dropout_p > 0.0) {
        dm = gqa_output_reshape(dm);
      }
    }
    ctx->save_for_backward({query_n, key_n, value_n, P, dm, out});
    return out;
  }

  static std::vector<at::Tensor> backward(
      torch::autograd::AutogradContext* ctx,
      torch::autograd::variable_list grad_output) {
    PT_EAGER_TRACE;
    torch::autograd::variable_list saved_vars = ctx->get_saved_variables();
    auto grad_out = grad_output[0];
    auto query = saved_vars[0];
    auto key = saved_vars[1];
    auto value = saved_vars[2];
    auto P = saved_vars[3];
    auto dm = saved_vars[4];
    auto fwd_out = saved_vars[5];
    auto enable_gqa = ctx->saved_data["enable_gqa"].toBool();

    if (enable_gqa) {
      grad_out = gqa_input_reshape_bwd(query, value, grad_out);
      fwd_out = gqa_input_reshape_bwd(query, value, fwd_out);
    }

    auto output = dispatch_sdpa_bwd_wrap(
        grad_out,
        query,
        key,
        value,
        P,
        dm,
        ctx->saved_data["is_causal"].toBool(),
        ctx->saved_data["dropout_p"].toScalar().toFloat(),
        ctx->saved_data["scale"].toScalar().toFloat(),
        fwd_out);
    auto query_grad = std::get<0>(output);
    auto key_grad = std::get<1>(output);
    auto value_grad = std::get<2>(output);
    if (enable_gqa) {
      query_grad = gqa_output_reshape(query_grad);
      key_grad = gqa_output_reshape(key_grad);
      value_grad = gqa_output_reshape(value_grad);
    }

    return {
        query_grad,
        key_grad,
        value_grad,
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor()};
  }
};

at::Tensor fused_sdpa_autograd_wrap(
    const at::Tensor& query,
    const at::Tensor& key,
    const at::Tensor& value,
    const ::std::optional<at::Tensor>& attn_mask,
    double dropout_p,
    bool is_causal,
    ::std::optional<double> scale,
    bool enable_gqa) {
  PT_EAGER_TRACE;
  return FusedSDPAAutogradHPU::apply(
      query, key, value, attn_mask, dropout_p, is_causal, scale, enable_gqa);
}

// When below flag is enabled, aten.scaled_dot_product_attention is overridden
// in torch.compile and eager
static const bool OVERRIDE_FSDPA =
    GET_ENV_FLAG_NEW(PT_HPU_USE_OVERRIDE_ATEN_SDPA);

TORCH_LIBRARY_IMPL(aten, AutogradHPU, m) {
  if (OVERRIDE_FSDPA) {
    m.impl("scaled_dot_product_attention", fused_sdpa_autograd_wrap);
  }
}

TORCH_LIBRARY_IMPL(aten, HPU, m) {
  m.impl("_fused_sdp_choice", fused_sdp_choice_hpu);
}

} // namespace eager
} // namespace habana
